src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/TableOfContents.java

```diff
@@ -0,0 +1,108 @@
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package jdk.javadoc.internal.doclets.formats.html;
+
+import jdk.javadoc.internal.doclets.formats.html.markup.Entity;
+import jdk.javadoc.internal.doclets.formats.html.markup.HtmlAttr;
+import jdk.javadoc.internal.doclets.formats.html.markup.HtmlId;
+import jdk.javadoc.internal.doclets.formats.html.markup.HtmlStyle;
+import jdk.javadoc.internal.doclets.formats.html.markup.HtmlTree;
+import jdk.javadoc.internal.doclets.formats.html.markup.ListBuilder;
+import jdk.javadoc.internal.doclets.formats.html.markup.TagName;
+
+/**
+ * A class used by various {@link HtmlDocletWriter} subclasses to build tables of contents.
+ */
+public class TableOfContents {
+    private final ListBuilder listBuilder;
+    private final HtmlDocletWriter writer;
+
+    /**
+     * Constructor
+     * @param writer the writer
+     */
+    public TableOfContents(HtmlDocletWriter writer) {
+        this.writer = writer;
+        listBuilder = new ListBuilder(HtmlTree.OL(HtmlStyle.tocList));
+    }
+
+    /**
+     * Adds a link to the table of contents.
+     * @param id the link fragment
+     * @param label the link label
+     * @return this object
+     */
+    public TableOfContents addLink(HtmlId id, Content label) {
+        listBuilder.add(writer.links.createLink(id, label).put(HtmlAttr.TABINDEX, "0"));
+        return this;
+    }
+
+    /**
+     * Adds a new nested list to add new items to.
+     */
+    public void pushNestedList() {
+        listBuilder.pushNestedList(HtmlTree.OL(HtmlStyle.tocList));
+    }
+
+    /**
+     * Closes the current nested list and go back to the parent list.
+     */
+    public void popNestedList() {
+        listBuilder.popNestedList();
+    }
+
+    /**
+     * Returns a sidebar content object containing a header with various controls as well as
+     * the given {@code content}.
+     *
+     * @param hasFilterInput whether to add a filter text input
+     * @return a content object
+     */
+    protected Content getSideBar(boolean hasFilterInput) {
```

Comment:
(Just asking)

In other builders for "complex" content, like `Table` and `Navigation` , we have methods like `toContent` or `getContent`.  Should we follow that naming convention?  While the name `getSidebar` is certainly more descriptive, is the returned object inherently a sidebar -- or is the fact that it is a sidebar just an artifact of how we currently choose to present the content?

Reply:
It is actually up to the CSS how the table of contents is rendered. I'm renaming the method to `toContent()`.